lib/git_worktree.rb

```diff
@@ -0,0 +1,336 @@
+                                 :tree      => merge_index.write_tree(@repo))
+  end
+
+  def merge_and_push(commit)
+    rebase = false
+    push_lock do
+      @saved_cid = @repo.ref(HEAD_REF).target.oid
+      merge(commit, rebase)
+      rebase = true
+      @repo.push(@remote_name, [HEAD_REF], :credentials => @cred)
+    end
+  end
+
+  def fetch
+    options = {:credentials => @cred}
+    ssl_no_verify_options(options) do
+      @repo.fetch(@remote_name, options)
+    end
+  end
+
+  def pull
+    lock { fetch_and_merge }
+  end
+
+  def fetch_and_merge
+    fetch
+    commit = @repo.ref("refs/remotes/#{@remote_name}/master").target
+    merge(commit)
+  end
+
+  def file_list(commit_sha = nil)
+    tree = lookup_commit_tree(commit_sha || @commit_sha)
+    return [] unless tree
+    tree.walk(:preorder).collect { |root, entry| "#{root}#{entry[:name]}" }
+  end
+
+  def find_entry(path, commit_sha = nil)
+    get_tree_entry(path, commit_sha)
+  end
+
+  def mv_file_with_new_contents(old_file, new_path, new_data, commit_sha = nil, default_entry_keys = {})
+    add(new_path, new_data, commit_sha, default_entry_keys)
+    remove(old_file, commit_sha)
+  end
+
+  def mv_file(old_file, new_file, commit_sha = nil)
+    entry = current_index[old_file]
+    return unless entry
+    entry[:path] = new_file
+    current_index(commit_sha).add(entry)
+    remove(old_file, commit_sha)
+  end
+
+  def mv_dir(old_dir, new_dir, commit_sha = nil)
+    old_dir = fix_path_mv(old_dir)
+    new_dir = fix_path_mv(new_dir)
+    updates = current_index(commit_sha).entries.select { |entry| entry[:path].start_with?(old_dir) }
+    updates.each do |entry|
+      entry[:path] = entry[:path].gsub(old_dir, new_dir)
+      current_index(commit_sha).add(entry)
+    end
+    current_index(commit_sha).remove_dir(old_dir)
+  end
+
+  private
+
+  def process_repo(options)
+    if options[:url]
+      clone(options[:url])
+    elsif options[:new]
+      create_repo
+    else
+      open_repo
+    end
+  end
+
+  def create_repo
+    @repo = @bare ? Rugged::Repository.init_at(@path, :bare) :
+                    Rugged::Repository.init_at(@path)
+    @repo.config['user.name']       = @name  if @name
+    @repo.config['user.email']      = @email if @email
+    @repo.config['merge.ff']        = 'only' if @fast_forward_merge
+  end
+
+  def open_repo
+    @repo = Rugged::Repository.new(@path)
+  end
+
+  def clone(url)
+    options = {:credentials => @cred, :bare => true, :remote => @remote_name}
+    ssl_no_verify_options(options) do
+      @repo = Rugged::Repository.clone_at(url, @path, options)
+    end
+  end
+
+  def fetch_entry(path, commit_sha = nil)
+    find_entry(path, commit_sha).tap do |entry|
+      raise MiqException::MiqGitEntryMissing, path unless entry
+    end
+  end
+
+  def fix_path_mv(dir_name)
+    dir_name = dir_name[1..-1] if dir_name[0] == '/'
+    dir_name += '/'            if dir_name[-1] != '/'
+    dir_name
```

Comment:
@Fryguy 
This is being used to manipulate directory names in the index rather than on the FileSystem.
The goal is to get a path that matches the way git stores its index entries without a leading slash and includes a trailing slash for directories.